File: src/BBox.cc
// This file is part of charto-store, copyright (C) 2016 BusFaster Ltd.
// Released under the MIT license, see LICENSE.

#include <cstdint>

#include "BBox.h"

using namespace charto;

template <typename Unit>
double BBox<Unit> :: sqDistTo(Unit x, Unit y) const {
	double dx = 0, dy = 0;

	if(x < sw.x) dx = x - sw.x;
	else if(x > ne.x) dx = x - ne.x;

	if(y < sw.y) dy = y - sw.y;
	else if(y >= ne.y) dy = y - ne.y;

	return(dx * dx + dy * dy);
}

/*
	The extended edges of the bounding box partition the coordinate system
	into 9 regions like this:

	ul / sw

	v               v
	1100 | 0100 | 0110
	-----+------+-----
	1000 | 0000 | 0010
	-----+------+-----
	1001 | 0001 | 0011<   lr / ne

	Result is a 4-bit code where each bit represents being on the outside of a
	particular edge. Points on edges are considered to belong to the region
	below or on the right side of them in this diagram.
*/

template <typename Unit>
unsigned int BBox<Unit> :: getRegion(Unit x, Unit y) const {
	return((
		((x < ul.x) << 3) |
		((y < ul.y) << 2) |
		((x < lr.x) << 1) |
		 (y < lr.y)
	) ^ 3);
}

// Integer version uses more bit twiddles to ensure it's branchless.

template <>
unsigned int BBox<uint32_t> :: getRegion(uint32_t x, uint32_t y) const {
	return((
		(((x - ul.x) >> 28) & 8) |
		(((y - ul.y) >> 29) & 4) |
		(((x - lr.x) >> 30) & 2) |
		 ((y - lr.y) >> 31)
	) ^ 3);
}

// Box edges or corners touching each other are considered not to intersect,
// because only the SW corner and edges touching it belong to each box.

template <typename Unit>
bool BBox<Unit> :: intersects(const BBox &other) const {
	return(
		   other.sw.x < ne.x
		&& other.sw.y < ne.y
		&& other.ne.x > sw.x
		&& other.ne.y > sw.y
	);
}

// Both bounding boxes must have nonzero width!
// Otherwise this may erroneously return true.

template <typename Unit>
bool BBox<Unit> :: containsX(const BBox &other) const {
	return(other.sw.x >= sw.x && other.ne.x <= ne.x);
}

// Both bounding boxes must have nonzero height!
// Otherwise this may erroneously return true.

template <typename Unit>
bool BBox<Unit> :: containsY(const BBox &other) const {
	return(other.sw.y >= sw.y && other.ne.y <= ne.y);
}

// Both bounding boxes must have nonzero width and height!
// Otherwise this may erroneously return true.

template <typename Unit>
bool BBox<Unit> :: contains(const BBox &other) const {
	return(containsX(other) && containsY(other));
}

template <typename Unit>
void BBox<Unit> :: extend(Unit x, Unit y) {
	if(x < sw.x) sw.x = x;
	if(y < sw.y) sw.y = y;
	if(x > ne.x) ne.x = x;
	if(y > ne.y) ne.y = y;
}

template class charto::BBox<double>;
template class charto::BBox<uint32_t>;

#include "nbind/nbind.h"

using namespace nbind;

NBIND_CLASS(BBox<double>, BBox) {
	construct<double, double, double, double>();

	getter(getSW, "sw");
	getter(getNE, "ne");

//	method(sqDistTo);
	multimethod(getRegion, args(double, double));

	method(intersects);
	multimethod(contains, args(const BBox<double> &));

	multimethod(extend, args(double, double));
}

NBIND_CLASS(BBox<uint32_t>, BBoxInt) {
	construct<uint32_t, uint32_t, uint32_t, uint32_t>();

	getter(getSW, "sw");
	getter(getNE, "ne");

//	method(sqDistTo);
	multimethod(getRegion, args(uint32_t, uint32_t));

	method(intersects);

	multimethod(extend, args(uint32_t, uint32_t));
}

File: src-ts/BBox.ts
// This file is part of charto-store, copyright (C) 2016 BusFaster Ltd.
// Released under the MIT license, see LICENSE.

import {Coord} from './Coord';

function assert(condition: Boolean) { if(!condition) throw(new Error('Assert failed')); }

export class BBox {

	constructor(sw: Coord, ne?: Coord);
	constructor(x1: number, y1: number);
	constructor(x1: number, y1: number, x2: number, y2: number);

	constructor(a: number | Coord, b?: number | Coord, x2?: number, y2?: number) {
		let x1: number;
		let y1: number;

		if(typeof(a) == 'number') {
			assert(typeof(b) == 'number');

			x1 = a;
			y1 = b as number;
		} else {
			({x: x1, y: y1} = a);

			if(b) {
				assert(typeof(b) != 'number');

				({x: x2, y: y2} = b as Coord);
			}
		}

		if(!x2 && x2 !== 0) x2 = x1;
		if(!y2 && y2 !== 0) y2 = y1;

		assert(x1 <= x2 && y1 <= y2);

		this.x1 = x1;
		this.y1 = y1;
		this.x2 = x2!;
		this.y2 = y2!;
	}

	sw() { return(new Coord(this.x1, this.y1)); }
	ul = this.sw;

	ne() { return(new Coord(this.x2, this.y2)); }
	lr = this.ne;

	sqDistTo(xy: Coord): number;
	sqDistTo(x: number, y: number): number;

	sqDistTo(a: number | Coord, y?: number) {
		let x: number;
		let dx = 0, dy = 0;

		if(typeof(a) == 'number') x = a;
		else ({x, y} = a);

		if(x < this.x1) dx = x - this.x1;
		else if(x > this.x2) dx = x - this.x2;

		if(y < this.y1) dy = y - this.y1;
		else if(y >= this.y2) dy = y - this.y2;

		return(dx * dx + dy * dy);
	}

	area() {
		return((this.x2 - this.x1) * (this.y2 - this.y1));
	}

	getCenter() {
		return(new Coord((this.x1 + this.x2) / 2, (this.y1 + this.y2) / 2));
	}

	getRegion(xy: Coord): number;
	getRegion(x: number, y: number): number;

	getRegion(a: number | Coord, y?: number) {
		let x: number;

		if(typeof(a) == 'number') x = a;
		else ({x, y} = a);

		return((
			(((x - this.x1) >>> 28) & 8) |
			(((y - this.y1) >>> 29) & 4) |
			(((x - this.x2) >>> 30) & 2) |
			 ((y - this.y2) >>> 31)
		) ^ 3);
	}

	/** Box edges or corners touching each other are considered not to intersect,
	  * because only the SW corner and edges touching it belong to each box. */

	intersects(other: BBox) {
		return(
			   other.x1 < this.x2
			&& other.y1 < this.y2
			&& other.x2 > this.x1
			&& other.y2 > this.y1
		);
	}

	containsX(other: BBox) {
		return(other.x1 >= this.x1 && other.x2 <= this.x2);
	}

	containsY(other: BBox) {
		return(other.y1 >= this.y1 && other.y2 <= this.y2);
	}

	contains(other: BBox): boolean;
	contains(xy: Coord): boolean;
	contains(x: number, y: number): boolean;

	contains(a: number | Coord | BBox, y?: number) {
		let x: number;

		if(typeof(a) == 'number') x = a;
		else if(a instanceof Coord) ({x, y} = a);
		else return(this.containsX(a) && this.containsY(a));

		return(this.getRegion(x, y!) == 0);
	}

	extend(xy: Coord): BBox;
	extend(x: number, y: number): BBox;

	extend(a: number | Coord, y?: number) {
		let x: number;

		if(typeof(a) == 'number') x = a;
		else ({x, y} = a);

		if(x < this.x1) this.x1 = x;
		if(y < this.y1) this.y1 = y!;
		if(x > this.x2) this.x2 = x;
		if(y > this.y2) this.y2 = y!;

		return(this);
	}

	x1: number;
	y1: number;

	x2: number;
	y2: number;

}
